### DISCUSSION TITLE
Enable UART console in LPCXpresso55s69 EVB flexcomm0 question

### DISCUSSION THREAD

**Discussion by MarsMSWu** [zephyrproject-rtos/zephyr#65498] [Q&A] [answered]
Hi,
  I had search some related topic(#55058 and #64084 ) for my question...
  I try to output debug log in my EVB by flexcomm0 and it always shows USAGE FAULT.
  Below is my _dts_ and _dtsi_ content for flexcomm0.
  In my understanding, if use flexcomm0 for debug console and it should be only configured in _*.dts_ and _*-pinctrl.dtsi_.
  Maybe add some clock setup/attach in _main(or init).c_ ... am I right ??

---------------------------------------------------------------------------------------------------

    aliases {
        usart-0 = &flexcomm0;
    };
    chosen {
        zephyr,console = &flexcomm0;
        zephyr,shell-uart = &flexcomm0;
    };
    &flexcomm0 {
        compatible = "nxp,lpc-usart";
        status = "okay";
        current-speed = <115200>;
        pinctrl-0 = <&pinmux_flexcomm0_usart>;
    pinctrl-names = "default";
    };

    &pinctrl {
	pinmux_flexcomm0_usart: pinmux_flexcomm0_usart {
            group0 {
                 pinmux = <FC0_TXD_SCL_MISO_WS_PIO0_30>,
                                 <FC0_RXD_SDA_MOSI_DATA_PIO0_29>;
                 slew-rate = "standard";
            };
	};
    };
---------------------------------------------------------------------------------------------------
  Below is my init.c content, it do some clock attach/setup and these setting just refer to _zephyr/soc/arm/nxp_lpc/lpc55xxx/soc.c_

	/*!< Set up the clock sources */
	/*!< Configure FRO192M */
	/*!< Ensure FRO is on  */
	POWER_DisablePD(kPDRUNCFG_PD_FRO192M);
	/*!< Set up FRO to the 12 MHz, just for sure */
	CLOCK_SetupFROClocking(12000000U);
	/*!< Switch to FRO 12MHz first to ensure we can change the clock */
	CLOCK_AttachClk(kFRO12M_to_MAIN_CLK);
	/* Enable FRO HF(96MHz) output */
	CLOCK_SetupFROClocking(96000000U);
	/*!< Set up dividers */
	CLOCK_SetClkDiv(kCLOCK_DivAhbClk, 1U, false);
	/*!< Set up clock selectors - Attach clocks to the peripheries */
	CLOCK_AttachClk(kFRO_HF_to_MAIN_CLK);
	/* Enables the clock for the I/O controller.: Enable Clock. */
	CLOCK_EnableClock(kCLOCK_Iocon);
---------------------------------------------------------------------------------------------------
  I also add CONFIG in my _defconfig_ file, as below

	CONFIG_SOC_SERIES_LPC55XXX=y
	CONFIG_SOC_LPC55S69_CPU0=y
	CONFIG_CONSOLE=y
	CONFIG_UART_CONSOLE=y
	CONFIG_SERIAL=y
	CONFIG_UART_INTERRUPT_DRIVEN=y
	CONFIG_CORTEX_M_SYSTICK=y
	CONFIG_GPIO=y
	CONFIG_PINCTRL=y
---------------------------------------------------------------------------------------------------
But my putty always show USAGE FAULT
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/d83837df-8c28-4df0-8784-3835eacd138f)

Thanks.

**Comment by DerekSnell** [comment]
Hi @MarsMSWu ,
This eval board is already supported in Zephyr as the board `lpcxpresso55s69_cpu0`.  And I see this board already uses Flexcomm0 as the default UART for the Console and Shell.  You will find these chosen nodes in the [lpcxpresso55s69_cpu0.dts](https://github.com/zephyrproject-rtos/zephyr/blob/7be1a8119bfd43810aae71a04670e40c0cc4e04b/boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts#L43) devicetree file.  Therefore, you can build any sample or test, and that application will print out Flexcomm0, such as [hello_world](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/hello_world).  

The files that you mentioned, are these for a custom board?  I suggest you use this board in the repo  `lpcxpresso55s69_cpu0` as a reference, and compare your files with it.  If you create a new custom board, it is best to start with a working board already in the repo, like  `lpcxpresso55s69_cpu0`.  Clone that board folder, give it a new name, and rename the board within the cloned board files.  Ensure that clone builds and works with the original hardware.  And then you can start modifying the cloned board files to match the other hardware.

Let us know what you find.  Best regards

**Reply by MarsMSWu** [reply]
Hi @DerekSnell 
I had build hello_world for lpcxpresso55s69_cpu0, the console result is correct.
So, I had compare the zephyr.dts from build of hello_world and my project.
The compile result of device tree for flexcomm0 are the same, so maybe my setting is correct.

It should be something wrong...because I use my original project and add another board configure folder for lpc55s69...
I will keep study and if anyupdate then I'll feedback on web.

Do you know the related clock setting is need or not ?
I not found clock setting in sorce code of hello_world...maybe below setting is not need, right ????
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/a1dffe12-a284-40bf-9ca4-01fc2717d3b4)

Thanks.

**Reply by MarsMSWu** [reply]
Hi @DerekSnell ,
I had copy my board files to zephyr\boards\arm and compile hello_world by my board file.
The EVB can show related log, I think maybe my board configure is correct.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/41a99ac9-d174-4175-a52e-f007004420d4)

So, maybe should try to find why USAGE FAULT...then halting system, am I right ???

Thanks.

**Comment by DerekSnell** [comment]
Hi @MarsMSWu ,

Sorry, I am not sure what is causing the fault on your board.

I have a suggestion you can try, although I am not optimistic it will help.  I noticed in [clock_init()](https://github.com/zephyrproject-rtos/zephyr/blob/c07156ef3c70bd798ff5f6eb728594acf30c5d9d/soc/arm/nxp_lpc/lpc55xxx/soc.c#L81), the clock to Flexcomm0 is not initialized.  Testing on my LPCXpresso55S69 board, I see using the debugger that the clock tree is initialized already when the application starts executing.  The clock mux register `SYSCON->FCCLKSEL0` is set to 3 , which selects the FRO 96 MHz clock.  But this is not the hardware's default setting for that mux.  That tells me, the Boot ROM is configuring the clock for Flexcomm0 before the application executes.  Perhaps on your board, the Boot ROM is configured differently, and it is not initializing the clock for you.  You can try adding the clock init code for Flexcomm0 in `clock_init()`, and use this [Flexcomm2 clock code](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/nxp_lpc/lpc55xxx/soc.c#L179) as a reference.

If that change does not help, then I think the best option is to use a debugger with the startup code, and confirm what is causing the fault.  It seems you already tried building an image for your custom board, and then executed it on the LPCXpresso55S69 board, and that worked.  So it seems there is some hardware difference with your custom board related to this fault.  The fault capture you shared printed the PC was at 0x1000_F408 when the fault occurred, so that can help narrow down where in the code the fault occurs.  But it also seems this fault occurs every time for you, so it should be simple to step through the startup code and find where it occurs.  Since you also have code working on the eval board, you can use the debugger on that board to compare how it executes successfully, and what is different.  That should expose the root cause.

If a fault occurs the first time accessing a register of a peripheral, that is usually a bus fault caused by a disabled clock.  When the registers accessed are not clocked, this will lead to a bus fault.  If you find that is the case, you can view the registers in the debugger to confirm the clock configuration, and also compare it with the working eval board.  For example, `SYSCON->FCCLKSEL0` selects the clock source for Flexcomm3.  When set to 3 for the FRO 96 MHz clock, that FRO clock must also be enabled, which is done with the bit `ANACTRL->FRO192M_CTRL[ENA_96MHZCLK]`.

If you are not already using a debugger, one option is [NXP's MCUXpresso extension for VS Code](https://github.com/nxp-mcuxpresso/vscode-for-mcux/wiki).  We have [Zephyr Lab Guides](https://github.com/nxp-mcuxpresso/vscode-for-mcux/wiki/Training-Zephyr-Getting-Started-RT1060) posted there to help get started, including using the debugger to view peripheral registers.  If you have any questions or issues using this VS Code extension, please share them at the [GitHub Issues for MCUXpresso extension for VS Code](https://github.com/nxp-mcuxpresso/vscode-for-mcux/issues).

Please let us know what you find.  Best regards

**Reply by MarsMSWu** [reply]
Hi @DerekSnell ,
Thanks for your advice...USAGE FAULT error should dependon HW platform difference.
I remove all the clock setting in my init.c file, it means no any clock enable/attach for flexcomm0.
And I mask all my GPIO/ADC related setting in DTS file and try again.
It seems like the debug console can show in normally.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/c497563d-4ff7-4636-a151-0e30900e20b9)

Maybe the USAGE FAULT occurred by my DTS file, even if not used for control but still do initial for INPUT/OUTPUT or ACTIVE_H/L...
So, make my board something wrong...

Thanks.

**Comment by DerekSnell** [comment]
Thanks, @MarsMSWu 

Can you explain more about the clock settings you are doing in your init.c?  Typically, the [clock_init()](https://github.com/zephyrproject-rtos/zephyr/blob/c07156ef3c70bd798ff5f6eb728594acf30c5d9d/soc/arm/nxp_lpc/lpc55xxx/soc.c#L81) from soc.c will execute early in the initialization code, at the `PRE_KERNEL_1` level.  Is your init.c executing after that?  Is your fault occurring in the code you removed from init.c?  Or in other initialization code configured by your DTS changes?

Thanks

**Reply by MarsMSWu** [reply]
Hi @DerekSnell,
Because my lpc55s69 board file is modified from another board(rdrw610).
So, I think maybe GPIO/LED define impact the EVB then I modify the DTS file as below and test again.
(Just mask all related GPIO/ADC/LED configure in DTS file.)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/ae58c1ed-942b-4881-9ae0-c2efa92325fd)
After the GPIO/LED modified, the flexcomm0 debug console is work fine.

I mask the definition of LED and GPIO then the EVB can output console log.
But I'm not sure the clock setting for flexcomm0 is need or not, so I still remove clock setting from my init.c
(The init.c also had do SYS_INIT to register PRE_KERNEK_1 level.)
Currently, my init.c is no any clock setting for flexcomm0.
I'm not sure if change debug console to another flexcomm maybe FC6, need clock setting or not...

As below, my rdrw610 must do clock setting for flexcomm1 for console log output.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/1745660a-7457-4493-ac8b-6fc5f107d38a)
I don't know is that depend on chip or flexcomm difference or something.

Thanks.

**Comment by DerekSnell** [comment, marked as answer]
Hi @MarsMSWu ,

When you create a custom board, I recommend starting your board files from the closest existing board available using that SOC or SOC family.  In this case, I would use the devicetree files from the LPCXpresso55S69 board as your starting point, and then modify to match your hardware.  

I am sorry I caused the confusion about the Flexcomm0 clock.  That was a theory I had, but it now seems unrelated to your issue.  Because the Boot ROM enables the clock to Flexcomm0, the application is not required to configure it, unless the app wants to use different clock settings.  But it would not cause issues for the app to reconfigure the clock.  Just be sure the clock is configured before the Flexcomm driver is initialized.

And yes, other Flexcomms will need the clock enabled before initializing, like FC6.  When I tested the config after boot, I saw the Flexcomm0 and Flexcomm1 clocks were already enabled by the Boot ROM.  But most of the Flexcomm clocks are not enabled at boot, and the app needs to enable before using them.

Best regards

**Reply by MarsMSWu** [reply]
Hi @DerekSnell ,
It's very thanks for your help.
Yes, use the same family and modify to my HW board feature DTS is more suitable.
Exactly, I'm study how to add new one board file from previous project even if different chip family.
I'll take care to set clock_init while I got my project board and set FC6 to be debug console.

Thanks.
